src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs

```diff
@@ -0,0 +1,396 @@
+
+        public const string ImmutableListTypeName = "System.Collections.Immutable.ImmutableList";
+        public const string ImmutableListGenericTypeName = "System.Collections.Immutable.ImmutableList`1";
+        public const string ImmutableListGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableList`1";
+
+        public const string ImmutableStackTypeName = "System.Collections.Immutable.ImmutableStack";
+        public const string ImmutableStackGenericTypeName = "System.Collections.Immutable.ImmutableStack`1";
+        public const string ImmutableStackGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableStack`1";
+
+        public const string ImmutableQueueTypeName = "System.Collections.Immutable.ImmutableQueue";
+        public const string ImmutableQueueGenericTypeName = "System.Collections.Immutable.ImmutableQueue`1";
+        public const string ImmutableQueueGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableQueue`1";
+
+        public const string ImmutableSortedSetTypeName = "System.Collections.Immutable.ImmutableSortedSet";
+        public const string ImmutableSortedSetGenericTypeName = "System.Collections.Immutable.ImmutableSortedSet`1";
+
+        public const string ImmutableHashSetTypeName = "System.Collections.Immutable.ImmutableHashSet";
+        public const string ImmutableHashSetGenericTypeName = "System.Collections.Immutable.ImmutableHashSet`1";
+        public const string ImmutableSetGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableSet`1";
+
+        public const string ImmutableDictionaryTypeName = "System.Collections.Immutable.ImmutableDictionary";
+        public const string ImmutableDictionaryGenericTypeName = "System.Collections.Immutable.ImmutableDictionary`2";
+        public const string ImmutableDictionaryGenericInterfaceTypeName = "System.Collections.Immutable.IImmutableDictionary`2";
+
+        public const string ImmutableSortedDictionaryTypeName = "System.Collections.Immutable.ImmutableSortedDictionary";
+        public const string ImmutableSortedDictionaryGenericTypeName = "System.Collections.Immutable.ImmutableSortedDictionary`2";
+
+        // In the order we wish to detect a derived type.
+        private static readonly Type[] s_genericInterfacesWithAddMethods = new Type[]
+        {
+            typeof(IDictionary<,>),
+            typeof(IDictionary),
+            typeof(ISet<>),
+            typeof(IList<>),
+            typeof(ICollection<>),
+        };
+
+        // In the order we wish to detect a derived type.
+        private static readonly Type[] s_genericInterfacesWithoutAddMethods = new Type[]
+        {
+            typeof(IReadOnlyDictionary<,>),
+            typeof(IReadOnlyCollection<>),
+            typeof(IReadOnlyList<>),
+            typeof(IEnumerable<>),
+        };
+
+        public static Type GetImplementedCollectionType(Type queryType)
+        {
+            Debug.Assert(queryType != null);
```

Comment:
Could a check to queryType.IsAbstract and queryType.IsInterface help determine if a user-specified class should be used?

Reply:
Yes, I've added this check.